### ISSUE TITLE
QoL improvement: Withdraw money directly from bank when using commands

### ISSUE THREAD

**Issue by JTrolin** [LVPlayground/playground#589]
# QoL improvement: Withdraw money directly from bank when using commands

I propose a quality of life improvement when trying to use commands, when you're low on cash at hand. Instead of "forcing" players to withdraw some money to start a race, robbery, slap someone or similar, the money should be automatically used from their bank.

This should come with a convenience fee (as a set amount or percentage). Also needs to take into account whether or not the player is allowed to perform a `/withdraw` command at their current state (location, balance, what have you).

## Current behavior
1. Player has $0 cash on hand
2. Player tries to use `/slap`, which costs $500 and gets an error message
3. Player need to first withdraw $500
4. Player does `/slap` again

This is far from game breaking, but slightly annoying. However many minor annoyances add up over time.

## Proposed behavior
1. Player has $0 cash on hand.
2. Player tries `/slap`
3. Money is automatically withdrawn, as if the player does `/withdraw` themselves.
4. A convenience fee is payed by the player.
5. Player slaps someone else.

## Request for Comments
I would appreciate feedback particularly on the whole `convenience fee` idea. I would argue that there is already an abundance of money, and most of the commands such as races, slaps etc costs in on the order of hundreds of dollars, while banks allow $5,354,228,880 to be stored.

1. **Should** there be a convenience fee?
2. Should it be a flat amount? E.g. $100 per withdraw, regardless of amount withdrawn?
3. Should it be a percentage, of say +50% withdrawn is paid as a fee? So $500 withdrawn would add a $250 fee.
4. If it is a percentage, what is a reasonable amount?
5. Are there any commands that should be excluded from this, such as buying properties for example?

**Comment by SophiaLVP** [comment]
This is a nice suggestion JT. I do like to share my opinion on this one if you like
1. Your proposed behavior for auto withdrawal of money for some commands like /slap, /minigames(whatsoever it is), /showmessage, /tow and /taxi, which returns no gain to the user in some manner can minimize a lot of stress and typing burden as you mentioned above. However, I don't want this thing to be implemented on situations like buying properties, buying weapons because it is a source of taking cash and there's a chance where people can rob them through killing, which encourages deathmatch and money making scenerios. Players do want deathmatch and money side by side so it will encourage them to fight for it

2. I think of something odd for convenience fee you have mentioned above. For me, a new feature can play this role. We can introduce a concept of /customfee for LV bank perk which can be changed by the owner of the bank. /customfee [10-50] (where 10-50 are percentage factor) will happen on a player auto withdrawing amount when in hand cash is 0$. It will again encourage many players to make money out of it as so many players withdraw cash oftentimes and with auto withdrawl at 0$, the game will itself turn the tables.

That's what I think. What do you say about this?

**Comment by JTrolin** [comment]
I really appreciate you taking the time to think through this suggestion and leave your feedback Sophia!

1. I definitely agree with this. There should be an upper limit for how much may be automatically withdrawn. Definitely below $1,000,000, probably around the order of $50,000 to $100,000 in my opinion.

2. Just to clarify: you do agree that there should be some sort of fee-mechanic to withdrawing money from the bank and this discussion is more on how the implementation is done?

3. I definitely like your idea for implementing a fee, controllable by whomever owns the LVP Main Bank property. I believe this has been a feature previously. It feels very reasonable to tie these two things together.

Excellent feedback Sophia, thanks a lot again!

**Comment by RussellLVP** [comment]
Hi @JTrolin,

Thank you for the suggestion! I really like this :)

Some unsorted thoughts:

  * Could we make this a VIP-only feature? A strategy we could pursue is to focus on QoL improvements for VIPs.
  * I agree with both you and Sophia regarding not removing the need to carry larger sums of money. We should stimulate that.
  * I really like Sophia's suggestion of the _fee_ going to the owner of the LV Bank property, or at least part of it.

Implementation-wise, this should be fairly trivial to implement in the [FinanceRegulator](https://github.com/LVPlayground/playground/blob/master/javascript/features/finance/financial_regulator.js
) which is already in charge of both cash and bank account balances. It might require us to change checks all over the place from "_only take money when they carry sufficient cash_" to "_try to take money, and error if not possible_", however.

**Comment by SophiaLVP** [comment]
> * Could we make this a VIP-only feature? A strategy we could pursue is to focus on QoL improvements for VIPs.

Allow me to defend my point of view here. I do not like that as a VIP feature (but I think of something again for VIP's as well) and the reason is simple. Let everyone allow to auto withdraw money from their bank accounts. It will trigger the /customfee mechanism which only the owner of LV bank can set. If there's no owner then default fee percent should be 10. Regular players can withdraw directly up to $10000 and $50000 for VIP players (This can introduce as VIP perk) accordingly. In this way, everyone stays happy.

**Comment by Specifer** [comment]
This is simple yet big feature . I think we could stick to vip only feature . So for example VIP slaps quicky while a player has to withdraw some money and slap . There is nothing unequal in this , just takes more time . And there percentage shall be upto (10-35%) and not more than that , that's kind of barbaric .

**Comment by JTrolin** [comment]
- Regarding VIP limitation: I feel like we should implement the QoL improvement for all players, but as Sophia suggests consider some additional perk for VIPs. An increased maximum limit would be a good example for this.
- The fee I feel should be completely managed by the "LVP Main Bank" flag, and just apply whatever rules that flag does. If that puts all the money into the owners account, fine or if it just removes some money that's fine as well.

**Comment by SophiaLVP** [comment]
I am fully against making this feature for VIP players because the whole point of introducing /customfee will become pointless. We are trying to introduce a perk where owner can collect a fees for auto withdrawal of money and reserving it just for VIP players will be quite unfair. We are looking at balancing economical as well as player status values so I think its better not to be given away this whole thing to the VIPs. But as I have explained above in my recent comment, we will give perks to VIPs which will be not out of the way.

**Comment by Specifer** [comment]
You said it already in the above comment, no need to repeat. JT is also against VIP only but We have to hear others thinking on this, don't we? Because for other commands, there is a need and without it players can't use the whole tree feature, for example account alias. But this, this is nothing but a shortcut by not WITHDRAWing. There will not be players who are unhappy about this.

**Comment by SophiaLVP** [comment]
I strongly disagree with you there @Specifer. You need to look at the both sides but as far as I am seeing, your observations are quite one sided (mostly towards vips). I welcome your opinion so do the others.

**Comment by Specifer** [comment]
> You said it already in the above comment , no need to repeat. JT is also against VIP only but We have to hear others thinking on this. 


I am all hands for players too but that makes this thing more underrated .

>I welcome your opinion and so do others.

What does that mean? Everyone's opinion is to be considered

**Comment by OttoRocket-LVP** [comment]
First of all:
- I love the idea. I sometimes have that I want to enter a minigame but have no money. Then after withdrawing it I'm too late to join the minigame.

- I also like the idea of having a property where the money can go to.

- And I also agree we should stimulate players to run around with cash.


Last, I do think such QoL features are good for VIPs, they're not really necessary but nice to have, thus a reason to donate. I don't think it renders the property useless as we have quite some VIPs around.
